pkg/portal/cluster/machines.go

```diff
@@ -0,0 +1,87 @@
+package cluster
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+
+	machinev1beta1 "github.com/openshift/api/machine/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type MachinesInformation struct {
+	Name              string `json:"name"`
+	CreatedTime       string `json:"createdTime"`
+	LastUpdated       string `json:"lastUpdated"`
+	ErrorReason       string `json:"errorReason"`
+	ErrorMessage      string `json:"errorMessage"`
+	LastOperation     string `json:"lastOperation"`
+	LastOperationDate string `json:"lastOperationDate"`
+	Status            string `json:"status"`
+}
+
+type MachineListInformation struct {
+	Machines []MachinesInformation `json:"machines"`
+}
+
+func MachinesFromMachineList(machines *machinev1beta1.MachineList) *MachineListInformation {
+	final := &MachineListInformation{
+		Machines: make([]MachinesInformation, 0, len(machines.Items)),
+	}
+
+	for _, machine := range machines.Items {
+		lastOperation := "Unknown"
+		if machine.Status.LastOperation != nil &&
+			machine.Status.LastOperation.Description != nil {
+			lastOperation = *machine.Status.LastOperation.Description
+		}
+
+		lastOperationDate := "Unknown"
+		if machine.Status.LastOperation != nil &&
+			machine.Status.LastOperation.LastUpdated != nil {
+			lastOperationDate = machine.Status.LastOperation.LastUpdated.String()
+		}
+
+		status := "Unknown"
+		if machine.Status.Phase != nil {
+			status = *machine.Status.Phase
+		}
+
+		errorReason := "None"
+		if machine.Status.ErrorReason != nil {
+			errorReason = string(*machine.Status.ErrorReason)
+		}
+
+		errorMessage := "None"
+		if machine.Status.ErrorMessage != nil {
+			errorMessage = *machine.Status.ErrorMessage
+		}
+
+		final.Machines = append(final.Machines, MachinesInformation{
+			Name:              machine.Name,
+			CreatedTime:       machine.CreationTimestamp.String(),
+			LastUpdated:       machine.Status.LastUpdated.String(),
+			ErrorReason:       errorReason,
+			ErrorMessage:      errorMessage,
+			LastOperation:     lastOperation,
+			LastOperationDate: lastOperationDate,
+			Status:            status,
+		})
+	}
+
+	return final
```

Comment:
The single function is not helping as there is an issue while converting the Status.Something to *string.
I think the code is more clear the way it is right now. If you want I can still create those multiple function but still I think the code is clear the way it is right now. Let me know your thoughts.

Reply:
@s-amann 's suggestion was even better as it applied DRY, but if it can't be applied, no more debate.

Regarding a single long function, IMHO it is always cleaner/better to have a short function that gives you an overall idea about what the code does without going further into the details. I think good encapsulation/high cohesion starts with having short functions. 

The clear advantage I see with the multiple functions I suggested is that you can understand what "MachinesFromMachineList" does reading just 18 LOC (the function itself) and later you can go read each of the narrow functions it calls to understand the details. I did "comment PR" instead of "suggest changes" as there is nothing wrong with the code and this is just a personal suggestion to improve it a bit (from my personal point of view). 

There is an interesting talk about [Clean Code functions](https://www.youtube.com/watch?v=QvrNwfuWbCQ) from Uncle Bob.

In the end, the decision is up to you 🙂

Reply:
Ohkay .. I am convinced :D , I will change the code to have those functions.